# **Armut Proje**

## **İş Problemi**

* Türkiye’nin en büyük online hizmet platformu olan Armut, hizmet verenler ile hizmet almak isteyenleri buluşturmaktadır.
* Bilgisayarın veya akıllı telefonunun üzerinden birkaç dokunuşla temizlik, tadilat, nakliyat gibi hizmetlere kolayca ulaşılmasını sağlamaktadır.
* Hizmet alan kullanıcıları ve bu kullanıcıların almış oldukları servis ve kategorileri içeren veri setini kullanarak Association Rule Learning ile ürün tavsiye sistemi oluşturulmak istenmektedir.

### **Veri Seti**

* Veri seti müşterilerin aldıkları servislerden ve bu servislerin kategorilerinden oluşmaktadır.
* Alınan her hizmetin tarih ve saat bilgisini içermektedir.

$\star$ **UserId:** Müşteri numarası

$\star$ **ServiceId:** Her kategoriye ait anonimleştirilmiş servislerdir. (Örnek : Temizlik kategorisi altında koltuk yıkama servisi) Bir ServiceId farklı kategoriler altında bulanabilir ve farklı kategoriler altında farklı servisleri ifade eder.
(Örnek: CategoryId’si 7 ServiceId’si 4 olan hizmet petek temizliği iken CategoryId’si 2 ServiceId’si 4 olan hizmet mobilya montaj)

$\star$ **CategoryId:** Anonimleştirilmiş kategorilerdir. (Örnek : Temizlik, nakliyat, tadilat kategorisi)

$\star$ **CreateDate:** Hizmetin satın alındığı tarih

#### **GÖREV 1:** Veriyi Hazırlama

#### **Adım 1:** armut_data.csv dosyasınız okutunuz.

In [1]:
 !pip install mlxtend

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [2]:
import pandas as pd
pd.set_option('display.max_columns', None)
# pd.set_option('display.max_rows', None)
pd.set_option('display.width', 500)
# çıktının tek bir satırda olmasını sağlar.
pd.set_option('display.expand_frame_repr', False)
from mlxtend.frequent_patterns import apriori, association_rules

In [3]:
df = pd.read_csv("/content/drive/MyDrive/DSMLBC10/week_7 (10.11.22-16.11.22)/project/armut/armut_data.csv")
df.head()

UserId  ServiceId  CategoryId           CreateDate
0   25446          4           5  2017-08-06 16:11:00
1   22948         48           5  2017-08-06 16:12:00
2   10618          0           8  2017-08-06 16:13:00
3    7256          9           4  2017-08-06 16:14:00
4   25446         48           5  2017-08-06 16:16:00

In [4]:
df.shape

(162523, 4)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 162523 entries, 0 to 162522
Data columns (total 4 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   UserId      162523 non-null  int64 
 1   ServiceId   162523 non-null  int64 
 2   CategoryId  162523 non-null  int64 
 3   CreateDate  162523 non-null  object
dtypes: int64(3), object(1)
memory usage: 5.0+ MB


#### **Adım 2:** 

ServisID her bir CategoryID özelinde farklı bir hizmeti temsil etmektedir.

 ServiceID ve CategoryID'yi "_" ile birleştirerek hizmetleri temsil edecek yeni bir değişken oluşturunuz.

In [6]:
df["Hizmet"] = [str(row[1]) + "_" + str(row[2]) for row in df.values]
df.head()

UserId  ServiceId  CategoryId           CreateDate Hizmet
0   25446          4           5  2017-08-06 16:11:00    4_5
1   22948         48           5  2017-08-06 16:12:00   48_5
2   10618          0           8  2017-08-06 16:13:00    0_8
3    7256          9           4  2017-08-06 16:14:00    9_4
4   25446         48           5  2017-08-06 16:16:00   48_5

#### **Adım 3:** 

Veri seti hizmetlerin alındığı tarih ve saatten oluşmaktadır, herhangi bir sepet tanımı (fatura vb. ) bulunmamaktadır.

Association Rule Learning uygulayabilmek için bir sepet (fatura vb.) tanımı oluşturulması gerekmektedir.

 Burada sepet tanımı her bir müşterinin aylık aldığı hizmetlerdir. Örneğin; 7256 id'li müşteri 2017'in 8.ayında aldığı 9_4, 46_4 hizmetleri bir sepeti;
2017’in 10.ayında aldığı  9_4, 38_4  hizmetleri başka bir sepeti ifade etmektedir. Sepetleri unique bir ID ile tanımlanması gerekmektedir.

Bunun için öncelikle sadece yıl ve ay içeren yeni bir date değişkeni oluşturunuz. UserID ve yeni oluşturduğunuz date değişkenini "_" ile birleştirirek ID adında yeni bir değişkene atayınız.


In [7]:
import datetime as dt
df["CreateDate"] = df["CreateDate"].apply(pd.to_datetime)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 162523 entries, 0 to 162522
Data columns (total 5 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   UserId      162523 non-null  int64         
 1   ServiceId   162523 non-null  int64         
 2   CategoryId  162523 non-null  int64         
 3   CreateDate  162523 non-null  datetime64[ns]
 4   Hizmet      162523 non-null  object        
dtypes: datetime64[ns](1), int64(3), object(1)
memory usage: 6.2+ MB


In [8]:
df["New_Date"] = df["CreateDate"].dt.strftime("%Y-%m")
df.head()

UserId  ServiceId  CategoryId          CreateDate Hizmet New_Date
0   25446          4           5 2017-08-06 16:11:00    4_5  2017-08
1   22948         48           5 2017-08-06 16:12:00   48_5  2017-08
2   10618          0           8 2017-08-06 16:13:00    0_8  2017-08
3    7256          9           4 2017-08-06 16:14:00    9_4  2017-08
4   25446         48           5 2017-08-06 16:16:00   48_5  2017-08

In [9]:
df["SepetID"] = [str(row[0]) + "_" + str(row[5]) for row in df.values]
df.head()

UserId  ServiceId  CategoryId          CreateDate Hizmet New_Date        SepetID
0   25446          4           5 2017-08-06 16:11:00    4_5  2017-08  25446_2017-08
1   22948         48           5 2017-08-06 16:12:00   48_5  2017-08  22948_2017-08
2   10618          0           8 2017-08-06 16:13:00    0_8  2017-08  10618_2017-08
3    7256          9           4 2017-08-06 16:14:00    9_4  2017-08   7256_2017-08
4   25446         48           5 2017-08-06 16:16:00   48_5  2017-08  25446_2017-08

#### **GÖREV 2:** Birliktelik Kuralları Üretiniz.

#### **Adım 1:** Aşağıdaki gibi sepet hizmet pivot table’i oluşturunuz.

In [11]:
df_pt = df.groupby(["SepetID","Hizmet"])["CategoryId"].count().unstack().fillna(0).applymap(lambda x: 1 if x>0 else 0)
df_pt

Hizmet         0_8  10_9  11_11  12_7  13_11  14_7  15_1  16_8  17_5  18_4  19_6  1_4  20_5  21_5  22_0  23_10  24_10  25_0  26_7  27_7  28_4  29_0  2_0  30_2  31_6  32_4  33_4  34_6  35_11  36_1  37_0  38_4  39_10  3_5  40_8  41_3  42_1  43_2  44_0  45_6  46_4  47_7  48_5  49_1  4_5  5_11  6_7  7_3  8_5  9_4
SepetID                                                                                                                                                                                                                                                                                                               
0_2017-08        0     0      0     0      0     0     0     0     0     0     0    0     0     0     0      0      0     0     0     0     0     0    0     0     0     0     0     0      0     0     0     0      0    0     0     0     0     0     0     0     1     0     1     0    0     0    0    0    0    0
0_2017-09        0     0      0     0      0     0     0     0     0     0     0    0     0     0     0      0      0     0     0     0     0     0    0     0     0     0     0     0      0     0     0     0      0    0     0     0     0     0     0     0     0     0     1     0    1     0    0    0    0    0
0_2018-01        0     0      0     0      0     0     0     0     0     0     0    0     0     0     0      0      0     0     0     0     0     0    0     1     0     0     0     0      0     0     0     0      0    0     0     0     0     1     0     0     0     0     0     0    0     0    0    1    0    0
0_2018-04        0     0      0     0      0     1     0     0     0     0     0    0     0     0     0      0      0     0     0     0     0     0    0     1     0     0     0     0      0     0     0     0      0    0     0     0     0     1     0     0     0     0     0     0    0     0    0    0    0    0
10000_2017-08    0     0      0     0      0     0     0     0     0     0     0    0     0     0     0      0      0     0     0     0     0     0    0     0     0     0     0     0      0     0     0     0      0    0     0     0     0     0     0     0     1     0     0     0    0     0    0    0    0    0
...            ...   ...    ...   ...    ...   ...   ...   ...   ...   ...   ...  ...   ...   ...   ...    ...    ...   ...   ...   ...   ...   ...  ...   ...   ...   ...   ...   ...    ...   ...   ...   ...    ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...  ...  ...  ...  ...
99_2017-12       1     0      0     0      0     0     0     0     0     0     0    0     1     0     0      0      0     0     0     0     0     0    0     0     0     0     0     0      0     0     0     0      0    0     0     0     0     0     0     0     0     0     0     0    0     0    0    0    0    0
99_2018-01       1     0      0     0      0     0     0     0     0     0     0    0     1     0     0      0      0     0     0     0     0     0    0     0     0     0     0     0      0     0     0     0      0    0     0     0     0     0     0     0     0     0     0     0    0     0    0    0    0    0
99_2018-02       1     0      0     0      0     0     0     0     0     0     0    0     0     0     0      0      0     0     0     0     0     0    0     0     0     0     0     0      0     0     0     0      0    0     0     0     0     0     0     0     0     0     0     0    0     0    0    0    0    0
9_2018-03        0     0      0     0      0     0     0     0     0     0     0    0     0     0     0      0      0     1     0     0     0     0    0     0     0     0     0     0      0     0     0     0      0    0     0     0     0     0     0     0     0     0     0     0    0     0    0    0    0    0
9_2018-04        0     0      0     0      0     0     0     0     0     0     1    0     0     0     0      0      0     1     0     0     0     0    0     0     0     0     0     0      0     0     0     0      0    0     0     0     0     0     0     0     0     0 

#### **Adım 2:** Birliktelik kurallarını oluşturunuz.

In [13]:
frequent_itemsets = apriori(df_,
                            min_support=0.01,
                            use_colnames=True)
frequent_itemsets.sort_values("support", ascending=False).head()

support itemsets
8   0.238121   (18_4)
19  0.130286    (2_0)
5   0.120963   (15_1)
39  0.067762   (49_1)
28  0.066568   (38_4)

In [14]:
rules = association_rules(frequent_itemsets,
                          metric="support",
                          min_threshold=0.01)

In [17]:
rules[(rules["support"]>0.01) & (rules["confidence"]>0.1) & (rules["lift"]>2)] 

antecedents consequents  antecedent support  consequent support   support  confidence      lift  leverage  conviction
2        (2_0)      (15_1)            0.130286            0.120963  0.033951    0.260588  2.154278  0.018191    1.188833
3       (15_1)       (2_0)            0.120963            0.130286  0.033951    0.280673  2.154278  0.018191    1.209066
4       (33_4)      (15_1)            0.027310            0.120963  0.011233    0.411311  3.400299  0.007929    1.493211
10      (22_0)      (25_0)            0.047515            0.042895  0.011120    0.234043  5.456141  0.009082    1.249553
11      (25_0)      (22_0)            0.042895            0.047515  0.011120    0.259247  5.456141  0.009082    1.285834
12       (2_0)      (22_0)            0.130286            0.047515  0.016568    0.127169  2.676409  0.010378    1.091260
13      (22_0)       (2_0)            0.047515            0.130286  0.016568    0.348700  2.676409  0.010378    1.335350
14       (2_0)      (25_0)            0.130286            0.042895  0.013437    0.103136  2.404371  0.007849    1.067168
15      (25_0)       (2_0)            0.042895            0.130286  0.013437    0.313257  2.404371  0.007849    1.266432
18      (38_4)       (9_4)            0.066568            0.041393  0.010067    0.151234  3.653623  0.007312    1.129413
19       (9_4)      (38_4)            0.041393            0.066568  0.010067    0.243216  3.653623  0.007312    1.233418

#### **Adım 3:** arl_recommender fonksiyonunu kullanarak en son 2_0 hizmetini alan bir kullanıcıya hizmet önerisinde bulununuz.


In [18]:
def arl_recommender(rules_df, product_id, rec_count=1):
    sorted_rules = rules_df.sort_values("lift", ascending=False)
    recommendation_list = []
    for i, product in sorted_rules["antecedents"].items():
        for j in list(product):
            if j == product_id:
                recommendation_list.append(list(sorted_rules.iloc[i]["consequents"]))
    return recommendation_list[:rec_count]
    
arl_recommender(rules, "2_0", 5)

[['2_0'], ['15_1'], ['9_4'], ['25_0'], ['38_4']]